### PR TITLE
str_format: always return length of written string

### PR DESCRIPTION
**Pull request by def-**
As deducted from https://github.com/ddnet/ddnet/pull/1527

### CODE REVIEW

File: src/base/system.h [hunk 1]
```diff
@@ -1019,7 +1019,7 @@ int str_length(const char *str);
 		... - Parameters for the formatting.
 
 	Returns:
-		Length of written string
+		Length of written string, negative value on error
```
Comment (heinrich5991) [inline src/base/system.h]:
```suggestion
		Length of written string
```

Comment (def-) [reply src/base/system.h]: Why?

**Comment by def-** [comment]
Looks like we have to change the behaviour for Windows too:

> Both _vsnprintf and _vsnwprintf functions return the number of characters written if the number of characters to write is less than or equal to count; if the number of characters to write is greater than count, these functions return -1 indicating that output has been truncated.

File: src/base/system.h (hunk 1 above)
Comment (def-) [reply src/base/system.h]: It can still fail, for example on parsing the format string and will return a negative value then.

**Comment by def-** [comment]
I propose to simply use vsnprintf on Windows too, which behaves as expected:

> The vsnprintf function returns the number of characters written, not counting the terminating null character. If the buffer size specified by count is not sufficiently large to contain the output specified by format and argptr, the return value of vsnprintf is the number of characters that would be written, not counting the null character, if count were sufficiently large. If the return value is greater than count - 1, the output has been truncated. A return value of -1 indicates that an encoding error has occurred.

https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/vsnprintf-vsnprintf-vsnprintf-l-vsnwprintf-vsnwprintf-l?view=vs-2017

File: src/base/system.h (hunk 1 above)
Comment (heinrich5991) [reply src/base/system.h]: Mh. We give a compiler warning and thus an error. Still fair enough I guess?

Comment (def-) [reply src/base/system.h]: Only for static strings, someone might be crazy enough to build up the format string dynamically :D

Comment (heinrich5991) [reply src/base/system.h]: I turned on compiler warnings for that, too. :P No way to dynamically build format strings.

Comment (def-) [reply src/base/system.h]: Good.

**Comment by heinrich5991** [comment]
From the linked docs, I infer that `vsnprintf` only behaves that way since Windows 10.

I think we cannot rely on that just yet.

**Comment by def-** [comment]
Ok, new try...

**Comment by def-** [comment]
Microsoft's documentation is wrong btw. They claim:

> Both _vsnprintf and _vsnwprintf functions return the number of characters written if the number of characters to write is less than or equal to count; if the number of characters to write is greater than count, these functions return -1 indicating that output has been truncated.

But in truth _vsnprintf return `size` when truncated.

File: src/base/system.c [hunk 2]
```diff
@@ -2336,14 +2336,21 @@ int str_format(char *buffer, int buffer_size, const char *format, ...)
 	va_start(ap, format);
 	ret = _vsnprintf(buffer, buffer_size, format, ap);
 	va_end(ap);
+
+	buffer[buffer_size-1] = 0; /* assure null termination */
```
Comment (heinrich5991) [inline src/base/system.c]:
```suggestion
	buffer[buffer_size-1] = 0; /* assure null termination */
	if(ret < 0)
		ret = buffer_size - 1;
```

Comment (def-) [reply src/base/system.c]: Done

**Comment by heinrich5991** [comment]
Thanks for enduring my reviews. :)

bors r+

**Comment by def-** [comment]
Not your fault that Windows is insane.

**Pull request merged**